bpo-44376 - reduce pow() overhead for small exponents

tim-one opened the pull request:
BPO-44376
Reduce pow() overhead for small exponents.
Incidentally fixed a refcount oversight for
negative exponents.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-44376](https://bugs.python.org/issue44376) -->
https://bugs.python.org/issue44376
<!-- /issue-number -->

tim-one committed:
First stab.
Reduce pow() overhead for small exponents.
Incidentally fixed a refcount oversight for
negative exponents.

tim-one committed:
Repair that, e.g, pow(False, 1) returned False instead of 0.
It's annoying, but I don't want to fight about it ;-)

tim-one committed:
Fixed typo in comment.

mdickinson commented on Objects/longobject.c:
```diff
@@ -4185,6 +4185,7 @@ long_pow(PyObject *v, PyObject *w, PyObject *x)
                 goto Error;
             Py_DECREF(a);
             a = temp;
+            temp = NULL;
```
Urk! That one looks like my fault. We probably want to pull this into a separate PR, so that it can be backported. I'll aim to do that at some point this weekend.

tim-one replied:
Sounds good to me! Thank you. I removed the change from this PR.

tim-one committed:
Search for the most significant exponent bit right-to-left
instead of left-to-right. This cuts the time for exponent 2
by another 10%. Does that slow things for, e.g., an exponent
like 1 << 20? Sure, by a little.  But that exponent will go
on to do 20 squarings, and the extra native 32-bit shift-
and-tests in the loop are insignificant in comparison. For
a tiny exponent (like 2), the loop's shift-and-tests can
consume a significant part of the total time.

Also reverted the change I made to repair refcounting in
modular inverse code. Mark will make the change in a
distinct PR, and backport it. Without this change, the
"negative refcount" glitch stopped showing up in this PR's
code after I changed it to never return `a` directly.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into powdig

tim-one committed:
Skip the binary business entirely for exponents <= 3. It can't win
until the exponent is at least 4. Saves approaching another 10%
for exponent 2.

tim-one committed:
Make the bitmask variable of type `digit`. Using a giant signed int
insteads works OK, but was always a conceptual mess.

mdickinson reviewed (approved):
LGTM

tim-one committed:
And one more stab to reduce overhead for exponents <= 3.
Close to another 10% saved for exponent 2.

tim-one committed:
Repaired typo in comment.
